### PR TITLE
migration(monitors): Backfill MonitorCheckIns w/MonitorEnvironments

### LINKED ISSUE #45994: Crons: Backfill checkins with default environments

### PR DESCRIPTION
**Pull request by rjo100**
Backfills MonitorCheckIns with MonitorEnvironments. Dangerous migration. 

Follows work from https://github.com/getsentry/sentry/pull/45666

Closes https://github.com/getsentry/sentry/issues/45994

### CODE REVIEW

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    cursor = connection.cursor()
+    batch = []
+    for monitor_checkin_id, project_id, monitor_id, monitor_environment_id in queryset:
+        if monitor_environment_id:
+            continue
+
+        try:
+            monitor_environment = MonitorEnvironment.objects.filter(monitor_id=monitor_id)[0]
+        except IndexError:
```
Comment (rjo100) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:41]: this shouldn't happen but I'd like to check redash to confirm we don't have any naked `Monitor` with no associated `MonitorEnv`

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+def backfill_monitor_checkins(apps, schema_editor):
+    MonitorCheckIn = apps.get_model("sentry", "MonitorCheckIn")
+    MonitorEnvironment = apps.get_model("sentry", "MonitorEnvironment")
+
+    queryset = RangeQuerySetWrapperWithProgressBar(
+        MonitorCheckIn.objects.all().values_list(
+            "id",
+            "project_id",
+            "monitor",
+            "monitor_environment",
```
Comment (rjo100) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:28]: do these just pull the `id` of the object here? or the actual object?

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    )
+
+    cursor = connection.cursor()
+    batch = []
+    for monitor_checkin_id, project_id, monitor_id, monitor_environment_id in queryset:
+        if monitor_environment_id:
+            continue
+
+        try:
+            monitor_environment = MonitorEnvironment.objects.filter(monitor_id=monitor_id)[0]
```
Comment (wedamija) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:40]: I'd recommend running this query outside of the loop and using it to produce a dict of `monitor_id -> monitor_environment_id`. Then you can just use the `monitor_id` from the checkin to fetch the proper environment.

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/migrations/0384_backfill_monitor_env_checkins.py:28]: That will pull the whole `MonitorEnvironment` object. You can use `monitor_environment_id` here to just get the id. Do you actually need all of these fields though? I would've thought you'd only need `id`, `monitor_id` and `monitor_environment_id`

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+# Generated by Django 2.2.28 on 2023-03-14 23:45
+
+from django.db import connection, migrations
+from psycopg2.extras import execute_values
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+BATCH_SIZE = 100
+DEFAULT_ENVIRONMENT_NAME = "production"
```
Comment (wedamija) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:10]: Can remove this

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        ("sentry", "0383_mv_user_avatar"),
+    ]
+
+    operations = [
+        migrations.RunPython(
+            backfill_monitor_checkins,
+            migrations.RunPython.noop,
+            hints={
+                "tables": [
+                    "sentry_monitor" "sentry_monitorcheckin",
```
Comment (beezz) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:76]:
```suggestion
                    "sentry_monitor",
                    "sentry_monitorcheckin",
```

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py (hunk 2 above)
Comment (rjo100) [reply src/sentry/migrations/0384_backfill_monitor_env_checkins.py:28]: Right, I only want to pull the `id` haha. And yeah `project` left in accidentally

**Review by wedamija — COMMENTED** [review]
There's one case of a `Monitor` with two environments that I mentioned in slack, but other than that this should work ok.

File: src/sentry/migrations/0384_backfill_monitor_env_checkins.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        monitor_id: monitor_env_id
+        for monitor_id, monitor_env_id in MonitorEnvironment.objects.values_list("monitor_id", "id")
+    }
+
+    queryset = RangeQuerySetWrapperWithProgressBar(
+        MonitorCheckIn.objects.all().values_list(
+            "id",
+            "monitor_id",
+            "monitor_environment_id",
+        ),
```
Comment (rjo100) [inline src/sentry/migrations/0384_backfill_monitor_env_checkins.py:32]: I want to make sure that:
1. we only return monitorenvs with the `env` named `production`
2. in the case of multiple, return the most recent.

I'm not sure we can do 1 without selecting `environment` objects as well. 2 we can order by `date_added`

Comment (rjo100) [reply src/sentry/migrations/0384_backfill_monitor_env_checkins.py:32]: there's 2 instances of `1` in the db from one company

there's 1 instance of `2` from a monitor that moved projects

Comment (wedamija) [reply src/sentry/migrations/0384_backfill_monitor_env_checkins.py:32]: For 1. you can do `filter(environment__name="production")` when you build `monitor_mappings`. Ordering by `date_added` for 2 makes sense to me

Comment (rjo100) [reply src/sentry/migrations/0384_backfill_monitor_env_checkins.py]: Yeah I had 1 but my tests were failing. Realize now it was bad tests lol

**Review by wedamija — APPROVED** [review]

**Pull request merged**